### DISCUSSION TITLE
Source location relative test

### DISCUSSION THREAD

**Discussion by bilgili** [microsoft/STL#3943] [General] [answered]
We have found an issue with source_location where line numbers were reported wrong (https://developercommunity.visualstudio.com/t/std::source_location-started-reporting-w/10429484?q=source_location)
and I had a peek on the testing of the std::source_location.

It is as below:

```c++
const auto x = source_location::current();
    assert(x.line() == __LINE__ - 1);
``` 

It is relatively testing against the `__LINE__` macro. Problem can arise if somehow implementation-wise `__LINE__` is same as the `__builtin_LINE()` function. This will always report true. IMHO they should be static line numbers to test.

**Comment by cpplearner** [comment]
Agreed that static line numbers can be useful, but IMO they should be part of *compiler* tests. The library doesn't need to detect bugs with `__LINE__` (or `__builtin_LINE`), and even if such bugs are detected, there's no way to fix them in the library.

**Comment by bilgili** [comment]
I dont have a very strong opinion but I could catch the issue in my unit test which was checking a static line number. At least an additional static_assert may be beneficial. But I also agree compiler tests should definitely catch any regression.

**Comment by StephanTLavavej** [comment, marked as answer]
The preprocessor lives at a different "level" than `source_location`, so I think that this is fine for a library test.
